arborist/server.go

```diff
@@ -1304,13 +1304,49 @@ func (server *Server) handleUserRevokeAll(w http.ResponseWriter, r *http.Request
 func (server *Server) handleUserRevokePolicy(w http.ResponseWriter, r *http.Request) {
 	username := mux.Vars(r)["username"]
 	policyName := mux.Vars(r)["policyName"]
-	errResponse := revokeUserPolicy(server.db, username, policyName, getAuthZProvider(r))
-	if errResponse != nil {
-		errResponse.log.write(server.logger)
-		_ = errResponse.write(w, r)
+	authzProvider := getAuthZProvider(r)
+	policyInfo, err := fetchUserPolicyInfo(server.db, username, policyName)
+
+	if err != nil {
+		server.logger.Info("Error Fetching policy Info: %s", err.Error())
+		msg := fmt.Sprintf("Error Fetching policy Info: %s", err.Error())
+		response := newErrorResponse(msg, http.StatusInternalServerError, nil)
+		_ = response.write(w, r)
 		return
 	}
-	server.logger.Info("revoked policy %s for user %s", policyName, username)
+
+	if policyInfo != nil {
+
+		dbAuthzProvider := ""
+		providerExists := policyInfo.AuthzProvider.Valid
+		if providerExists {
+			dbAuthzProvider = policyInfo.AuthzProvider.String
+		}
+		server.logger.Debug("Policy - {name: %s, authz_provider: %s, expires_at: %s} assigned to user %s",
+			policyInfo.PolicyName, dbAuthzProvider, policyInfo.ExpiresAt, policyInfo.Username)
+
+		if !authzProvider.Valid || (providerExists && dbAuthzProvider == authzProvider.String) {
+			errResponse := revokeUserPolicy(
+				server.db, username, policyName, authzProvider)
+			if errResponse != nil {
+				errResponse.log.write(server.logger)
+				_ = errResponse.write(w, r)
+				return
+			}
+			server.logger.Info("revoked policy %s for user %s", policyName, username)
+		} else {
+			server.logger.Info("Cannot revoke policy `%s`. Authz_provider - `%s` and `%s` Mismatch",
```

Comment:
```suggestion
			server.logger.Info("Cannot revoke policy `%s`. Policy authz_provider `%s` and request authz_provider `%s` mismatch",
```
so we know which is which